File: graph/mst.cpp
#include<bits/stdc++.h>
using namespace std;

/*BLUE PRINT FOR GRAPH DATA STRUCTURE*/
class graphtype {

private:

	int source;
	int dest;
	int weight;
	graphtype *next;

	int getdest();
	void setdest(int);
	int getweight();
	void setweight(int);
	int getsource();
	void setsource(int);
	void setnext(graphtype*);
	graphtype* getnext();

public:

	void display_graph(graphtype *[], int );
	void input_graph(graphtype *[], int );
	void mst(graphtype *[], queue<int> ,int [], int);
	void find(int [],int , int );
	void setParent(int [], int);
	void showParent(int [],int );
};

void graphtype::setnext(graphtype *ptr)
{
	next = ptr;
}

graphtype* graphtype::getnext()
{
	return next;
}
void graphtype::setweight(int w)
{
	weight=w;
}

int graphtype::getweight()
{
	return weight;
}

void graphtype::setsource(int s)
{
	source=s;
}

int graphtype::getsource()
{
	return source;
}

void graphtype::setdest(int d)
{
	dest=d;
}

int graphtype::getdest()
{
	return dest;
}

void graphtype::input_graph(graphtype *vertice[], int total_edge)
{
	int i=1;
	while(total_edge--)
	{
		int s,d,w;
		//printf("enter source and destination of edege %d: ",i++);
		scanf("%d%d",&s,&d,&w);

		graphtype *ptr = new(graphtype);
		ptr->setdest(d);
		ptr->setnext(vertice[s]);
		vertice[s]=ptr;
	}
}

void graphtype::display_graph(graphtype *vertice[],int v)
{
	int e=1;

	for(int i=0; i<v; i++)
	{
		graphtype *ptr = vertice[i];
		printf("vertice[%d]--->",i);
		while(ptr!=NULL)
		{
			printf("%d--->",ptr->getdest());
			ptr=ptr->getnext();
			e++;
		}	printf("NULL\n");
	}
}

void graphtype::mst(graphtype *vertice[], queue<int> Queue,int parent[],int s)
{
	graphtype *ptr = vertice[s];

	printf("node processed id %d\n",s);

	while(ptr!=NULL)
	{
		int x = ptr->getdest();

		if(parent[x] == x)
		{
			parent[x] = s;
			Queue.push(x);
		}
		ptr=ptr->getnext();
	}

	//showParent(parent,6);

	if(!Queue.empty())
	{
		s = Queue.front();	Queue.pop();
		bfs(vertice,Queue,parent,s);
	}
}

void graphtype::setParent(int parent[], int v)
{
	for (int i = 0; i < v; ++i)
	{
		parent[i]=i;
	}
}

void graphtype::showParent(int parent[], int v)
{
	cout<<"\n";
	for (int i = 0; i < v; ++i)
		cout<<parent[i]<<" ";
	cout<<"\n";
}

void graphtype::find(int parent[], int s,int d)
{
	printf("shortest path : %d<--->",d);
	while(parent[d]!=s)
		printf("%d<--->",d=parent[d]);
	printf("%d\n",s);
}

int main()
{
	int v,e;
	cin>>v>>e;
	int parent[v],s,d;

	graphtype *vertice[v]={NULL},g;
	
	g.setParent(parent,v);
	g.input_graph(vertice,e);			//O(e)
	g.display_graph(vertice,v);		//O(v+e)

	queue<int> Queue;

	printf("Enter source and destination: ");
	scanf("%d%d",&s,&d);	parent[s]=-1;

	g.bfs(vertice,Queue,parent,s);
	g.find(parent,s,d);

	return 0;
}

File: graph/mst.py
#python program to reprsent graph in memory

class Graph :
	"""blue print for graph data structure"""
	def __init__(self,source = 0,dest = 0,weight = 0,next = None):
		self.source = source
		self.dest = dest
		self.weight = weight
		self.next = next

def find(parent,i):

	while i!=parent[i]:
		i = parent[i]
	return i

def union(parent,s,d):

	if(find(parent,s)!=find(parent,d)):
		return True
	else:
		return False

def mst(parent,front,v):

	weight,tmp,ptr = 0,None,front
	while v:

		s,d,w = ptr.source,ptr.dest,ptr.weight

		if union(parent,s,d):		#if no cycle formed by adding this egde then add this egde to mst
			parent[d] = s
			weight += w
			v-=1
			tmp = ptr
		else:
			tmp.next = ptr.next
		ptr = ptr.next
	
	tmp.next = None
	print("minimum weight of mst is {} ".format(weight))
	return front

def input_graph(front,rear,e):

	while e>0:
		e-=1
		#print("enter source destination and weight :",end = " ")
		s,d,w = input().split(" ")
		s,d,w = int(s),int(d),int(w)

		ptr = Graph(s,d,w)

		if(rear == None):
			front  = rear = ptr
		elif(w > (rear.weight)):
			rear.next = ptr
			rear = ptr
		elif(w < (front.weight)):
			ptr.next = front
			front = ptr
		else:
			tmp = front
			while(w > (tmp.next).weight):
				tmp = tmp.next
			ptr.next = tmp.next
			tmp.next = ptr
	return front

def display_graph(ptr):

	while(ptr):
		print("[{}|{}|{}]-->".format(ptr.source,ptr.dest,ptr.weight),end = " ")
		ptr = ptr.next
	print("NULL")

print("enter total number of vertices and edges ")

v,e = input().split(" ")
v,e = int(v),int(e)
parent = [0] * v
front = rear = None

for i in range(v):
	parent[i]=i

front = input_graph(front,rear,e)
display_graph(front)
front = mst(parent,front,v-1)
display_graph(front)